distributed/worker.py

```diff
@@ -2521,58 +2477,62 @@ async def execute(self, key, report=False):
             kwargs2 = pack_data(kwargs, data, key_types=(bytes, str))
             stop = time()
             if stop - start > 0.005:
-                self.startstops[key].append(
+                ts.startstops.append(
                     {"action": "disk-read", "start": start, "stop": stop}
                 )
                 if self.digests is not None:
                     self.digests["disk-load-duration"].add(stop - start)
 
             logger.debug(
-                "Execute key: %s worker: %s", key, self.address
+                "Execute key: %s worker: %s", ts.key, self.address
             )  # TODO: comment out?
+            assert key == ts.key
             try:
                 result = await self.executor_submit(
-                    key,
+                    ts.key,
                     apply_function,
                     args=(
                         function,
                         args2,
                         kwargs2,
                         self.execution_state,
-                        key,
+                        ts.key,
                         self.active_threads,
                         self.active_threads_lock,
                         self.scheduler_delay,
                     ),
                 )
+                # TODO: it seems insane that this needs to be here
+                # but it is occasionally cleared by something in executor_submit
+                self.executing.add(ts.key)
```

Comment:
Why does this need to be here?  What happens if it is removed?

Grepping through the codebase I see three paths for this:

1.  The task has been released during execution.  Maybe the user no longer needs the result
2.  The task is long-running, such as is called with `worker_client` or `secede`
3.  The same task is run twice (??)

In some of these situations I would expect the following `ts.state not in ("executing", "long-running")` check to clean things up.

Reply:
Hmm, 
I had put this in because there were tests in `test_failed_worker.py` that would throw assertion errors (but not fail the test in question) but I've just removed it and can no longer get those errors to show up, so it may have been due to a problem elsewhere that is now fixed.